server/src/main/kotlin/dev/robocode/tankroyale/server/core/NanoTimer.kt

```diff
@@ -1,58 +1,101 @@
+    private val jobExecuted = AtomicBoolean(false)
+    private var ready = AtomicBoolean(false)
+    private var pauseStartTime = AtomicLong(0L)
+    private var totalPauseDuration = AtomicLong(0L)
 
     /** Starts the timer. */
     fun start() {
-        thread = Thread { run() }
-        isRunning.set(true)
+        thread = Thread(this)
         thread?.start()
     }
 
     /** Stops the timer. */
-    fun stop() {
-        isRunning.set(false)
+    fun stop(): Boolean {
+        // This will prevent any other threads from executing the job, even if we haven't
+        val success = !jobExecuted.getAndSet(true)
         thread?.interrupt()
+        return success
     }
 
     /** Pauses the timer. */
     fun pause() {
-        isPaused.set(true)
+        pauseStartTime.compareAndSet(0L, System.nanoTime())
     }
 
     /** Resumes the timer after having been paused. */
     fun resume() {
-        isPaused.set(false)
+        val pauseStart = pauseStartTime.getAndSet(0L)
+        if (pauseStart == 0L) return
+
+        val diff = System.nanoTime() - pauseStart
+        totalPauseDuration.addAndGet(diff)
+
+        thread?.interrupt()
     }
 
-    /** Resets timer. */
-    fun reset() {
-        lastTime = System.nanoTime()
+    /** Notifies that the task is ready to be executed. */
+    fun notifyReady() {
+        ready.set(true)
+        thread?.interrupt()
     }
 
-    private fun run() {
-        while (isRunning.get() && !thread!!.isInterrupted) {
+    override fun run() {
+        var lastTime = System.nanoTime()
+        if (minPeriodInNanos > 0) {
+            while (true) {
+                val now = System.nanoTime()
+                val diff = lastTime + minPeriodInNanos + totalPauseDuration.get() - now
+                val isPaused = pauseStartTime.get() != 0L
+                if (diff <= 0 && !isPaused) break
+                try {
+                    Thread.sleep(if (isPaused) Duration.ofSeconds(1) else Duration.ofNanos(diff))
+                } catch (e: InterruptedException) {
+                    if (jobExecuted.get()) return
+                }
+            }
+        }
+
+        while (true) {
+            if (jobExecuted.get()) break
+
             val now = System.nanoTime()
-            if (now - lastTime >= periodInNanos) {
-                lastTime = now
-                if (!isPaused.get()) {
-                    job.run()
+            val diff = lastTime + maxPeriodInNanos + totalPauseDuration.get() - now
+            val isPaused = pauseStartTime.get() != 0L
+            if (ready.get() || diff <= 0 && !isPaused) {
+                executeJob()
+                break
+            } else {
+                try {
+                    Thread.sleep(if (isPaused) Duration.ofSeconds(1) else Duration.ofNanos(diff))
+                } catch (e: InterruptedException) {
+                    continue
                 }
             }
         }
+
         thread = null
     }
```

Comment:
I propose we split up the `run()` method into smaller methods to make it easier to read and to avoid redundant code as well following Clean Code principles. 🙂

```suggestion
    override fun run() {
        if (minPeriodInNanos <= 0) return

        waitForMinimumPeriod()
        waitAndExecuteJob()
        thread = null
    }

    private fun waitForMinimumPeriod() {
        val startTime = System.nanoTime()
        while (!canProceedAfterMinimumPeriod(startTime)) {
            sleepOrReturn()
        }
    }

    private fun waitAndExecuteJob() {
        val startTime = System.nanoTime()
        while (!jobExecuted.get()) {
            if (canExecuteJob(startTime)) {
                executeJob()
                break
            }
            sleep()
        }
    }

    private fun canProceedAfterMinimumPeriod(startTime: Long): Boolean {
        val now = System.nanoTime()
        val timeDiff = startTime + minPeriodInNanos + totalPauseDuration.get() - now
        return timeDiff <= 0 && !isPaused()
    }

    private fun canExecuteJob(startTime: Long): Boolean {
        val now = System.nanoTime()
        val timeDiff = startTime + maxPeriodInNanos + totalPauseDuration.get() - now
        return ready.get() || (timeDiff <= 0 && !isPaused())
    }

    private fun isPaused() = pauseStartTime.get() != 0L

    private fun sleepOrReturn() {
        try {
            if (isPaused()) {
                TimeUnit.SECONDS.sleep(1)
            } else {
                val sleepTime = calculateSleepTime(minPeriodInNanos)
                TimeUnit.NANOSECONDS.sleep(sleepTime)
            }
        } catch (e: InterruptedException) {
            if (jobExecuted.get()) return
        }
    }

    private fun sleep() {
        try {
            if (isPaused()) {
                TimeUnit.SECONDS.sleep(1)
            } else {
                val sleepTime = calculateSleepTime(maxPeriodInNanos)
                TimeUnit.NANOSECONDS.sleep(sleepTime)
            }
        } catch (ignore: InterruptedException) {
        }
    }

    private fun calculateSleepTime(periodInNanos: Long): Long {
        val now = System.nanoTime()
        val diff = periodInNanos + totalPauseDuration.get() - now
        return maxOf(diff, 0)
    }
````

Reply:
I'm all in favor of making the main loop easier to read, but I believe your refactoring introduces a few minor issues.

* The check `if (minPeriodInNanos <= 0)` should only skip `waitForMinimumPeriod`, not the rest of the main method.
* `waitForMinimumPeriod` and `waitAndExecuteJob` need to share the same `startTime`. I suppose we could move it into a member.
* If we run into the `if (jobExecuted.get()) return` in `waitForMinimumPeriod` we should abort the entire execution, not just this method.
* We have double the number of `System.nanoTime()` calls. Since we're not doing a busy wait anymore, this is probably not a big deal. However, nanoTime is one of the most expensive timing APIs since it uses performance counters on x86. That said, it is the only simple monotonic clock available in Java, so I wouldn't replace it either.

I'll repair the first two problems and incorporate your changes.

Reply:
You are free to fix anything you fell for - especially, if I introduced bugs in my refactoring suggestion. 👍